### ISSUE TITLE
Figure out libiconv shenanigans on macOS

### ISSUE THREAD

**Issue by Minoru** [newsboat/newsboat#1567]
I can't understand what version of libiconv Newsboat is linking to on macOS. I'd appreciate help from someone who has the hardware and knows what a "debugger", "linking", and "library search path" are.

# A bit of theory

Iconv is an API of three calls: `iconv_open`, `iconv`, and `iconv_close`. Sometimes the standard C library (libc) provides that (as is the case on Linux with glibc), sometimes it doesn't (as is the case on FreeBSD). To fill this gap, GNU provides a library called "libiconv".

The library can be built in two modes, which influence how functions are named: in the "plug" mode, they're prefixed with "lib", so `libiconv_open`, `libiconv`, `libiconv_close`. The other mode uses plain names (`iconv` etc.)

As far as I understand, on macOS we can have up to three iconv implementations:

1. in libc. Not sure if libc actually provides it; there's a [manpage](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/iconv_open.3.html) which mentions all of Linux, MacOS X, and iOS, so it's hard to tell;
2. in GNU libiconv shipped along with macOS;
3. in GNU libiconv installed from Homebrew or MacPorts.

The last two could be equivalent, or they could be different versions, or they could be built in different modes. This is largely based on [this StackOverflow question](https://stackoverflow.com/questions/57734434/libiconv-or-iconv-undefined-symbol-on-mac-osx).

# How Newsboat finds an iconv implementation to use

In C++, it's "simple": we `#include <iconv.h>`, that picks up the header for whichever library the OS views as preferable. The header takes care of the modes (it provides `#define iconv libiconv` and such to hide the name change from us). On Linux, this ends up being glibc; on other OSes, this is supposed to pick libiconv. On non-Linux, [we also add `-liconv` to our linking flags](https://github.com/newsboat/newsboat/blob/ae20c1e1e6ac240de443107a2639b2a7f841be4a/Makefile#L114-L117). This again means we're asking to be linked to libiconv.so.

For Rust this is slightly harder, because we can't import a C header. This means we have to provide our own function declarations, and, worst of all, take care of renaming:

https://github.com/newsboat/newsboat/blob/ae20c1e1e6ac240de443107a2639b2a7f841be4a/rust/libnewsboat/src/utils.rs#L893-L907

These `cfg_attr` things basically say "on FreeBSD, link with libiconv" and "on FreeBSD, link with this function as if it was called that".

# What's the problem?

Everything I described above passes all our tests on our continuous integration servers, which run Ubuntu, macOS with Homebrew, and FreeBSD.

However, our build instructions don't work for MacPorts (another package manager for macOS, basically). Newsboat maintainer in MacPorts [wrote patches](https://github.com/macports/macports-ports/pull/10355) which I'd like to include into Newsboat proper, but I can't since [they break Homebrew](https://github.com/newsboat/newsboat/pull/1557):

```
Undefined symbols for architecture x86_64:
  "_libiconv", referenced from:
      libnewsboat::utils::convert_text::hf7ddbcb26ca1a05c in libnewsboat.a(libnewsboat-d6bf53de572bf0da.libnewsboat.9m6dc9fb-cgu.6.rcgu.o)
  "_libiconv_close", referenced from:
      libnewsboat::utils::translit::h33e96ae162789cd9 in libnewsboat.a(libnewsboat-d6bf53de572bf0da.libnewsboat.9m6dc9fb-cgu.6.rcgu.o)
      libnewsboat::utils::convert_text::hf7ddbcb26ca1a05c in libnewsboat.a(libnewsboat-d6bf53de572bf0da.libnewsboat.9m6dc9fb-cgu.6.rcgu.o)
  "_libiconv_open", referenced from:
      libnewsboat::utils::translit::h33e96ae162789cd9 in libnewsboat.a(libnewsboat-d6bf53de572bf0da.libnewsboat.9m6dc9fb-cgu.6.rcgu.o)
      libnewsboat::utils::convert_text::hf7ddbcb26ca1a05c in libnewsboat.a(libnewsboat-d6bf53de572bf0da.libnewsboat.9m6dc9fb-cgu.6.rcgu.o)
```

This error means "Newsboat wants to use a function called `_libiconv`, but none of the libraries Newsboat links with have this function". (The underscore in front of the name is part of the so-called "calling conventions", pretend it's not there.)

# What can you do to help?

I suspect that this is caused by Newsboat trying to link with libiconv which is *not* compiled in the "plug" mode.

I'd like someone to examine the C compiler's include paths, examine what libraries are linked *without* my PR, and maybe write some small C programs to figure out what the correct incantation is to get them to link to the plug-libiconv. Once you got it linking, you might need to run _test/test_ in the debugger and check that it doesn't contain *both* `iconv` and `libiconv`.

The goal is to make this work (compile, link, pass tests) on: Linux, FreeBSD, macOS with Homebrew, and macOS with MacPorts. I expect you'll be experimenting with the latter locally; to test all others, just submit a draft PR and look at the CI's results.

(Bonus points if you can add MacPorts to our CI. Shouldn't be too hard given [we already have Homebrew instructions](https://github.com/newsboat/newsboat/blob/ae20c1e1e6ac240de443107a2639b2a7f841be4a/.cirrus.yml#L97)).

I realize I can probably get SSH into a macOS machine from Anka, or even buy one second-hand for cheap, but frankly, I'm just not *that* interested in macOS. There's only so much work I'm willing to put into supporting a platform I don't use, and macOS is reaching the end of its allowance. It's be great if some macOS hacker could shoulder some of the maintenance burden.

I'm of course open to answer questions regarding our build system, our code, and whatever else you might want to know — I don't expect you to just figure this all our for me. But you'll have to do the majority of the thinking and all the macOS testing.

**Comment by neelahaneesh03-ops** [comment]
Is this Issue still open? If it is, I'd like to work on this 🙋.

**Comment by Minoru** [comment]
Yup, it's still an open question. Note that our macOS CI job moved to GitHub Actions: https://github.com/newsboat/newsboat/blob/master/.github/workflows/macos.yml Other than that, it appears nothing changed.
